fdbclient/ManagementAPI.actor.cpp

```diff
@@ -2959,6 +2959,82 @@ ACTOR Future<Void> acknowledgeBulkLoadTask(Database cx, KeyRange range, UID task
 	return Void();
 }
 
+ACTOR Future<std::vector<RangeLockState>> _getUserRangeLockStates(Transaction* tr, KeyRange range) {
+	if (range.end > normalKeys.end) {
+		throw range_lock_failed();
+	}
+	// FIXME: multiple reads
+	state std::vector<RangeLockState> rangeLockStates;
+	RangeResult result = wait(krmGetRanges(tr, rangeLockPrefix, range));
+	for (int i = 0; i < result.size() - 1; i++) {
+		if (result[i].value.empty()) {
+			continue;
+		}
+		RangeLockState rangeLockState = decodeRangeLockState(result[i].value);
+		rangeLockStates.push_back(rangeLockState);
+	}
+	return rangeLockStates;
+}
+
+ACTOR Future<std::vector<RangeLockState>> getUserRangeLockStates(Database cx, KeyRange range) {
+	if (range.end > normalKeys.end) {
+		throw range_lock_failed();
+	}
+	state std::vector<RangeLockState> rangeLockStates;
+	loop {
+		rangeLockStates.clear();
+		state Transaction tr(cx);
+		try {
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			wait(store(rangeLockStates, _getUserRangeLockStates(&tr, range)));
+			break;
+		} catch (Error& e) {
+			wait(tr.onError(e));
+		}
+	}
+	return rangeLockStates;
+}
+
+ACTOR Future<Void> lockUserRange(Database cx, KeyRange range) {
```

Comment:
What's the rule of lock and unlock requests? If a client locks a range, who can unlock the range? The same client or any client can do this? If any client can unlock the range, how do we prevent accidental wrong unlocks? If it's the same client, do we need some sort of token stored in the DB so that the same client can use to unlock?

What happens for the following scenarios, i.e., overlapping locked ranges:
* `[d, f]` is locked, then `[a, g]` is requested;
* `[d, f]` is locked, then `[e, f]` is requested;
* `[d, f]` is locked, then `[e, g]` is requested;

Reply:
Any overlapping ranges with the same LockType will be merged